### DISCUSSION TITLE
How does bat expand and canonicalize UNIX paths?

### DISCUSSION THREAD

**Discussion by kirusfg** [sharkdp/bat#2215] [Q&A] [answered]
In a project of mine I have to deal with files provided by users, and those can be of the shortened form `~/path/to/some/file` or be a `/symlink/to/another/file`. 

While the `canonicalize` method from `std::fs` has been sufficient in resolving links, I am yet to find a proper way of expanding paths containing tildes (`~`). How does bat approach this?

**Comment by Enselic** [comment, marked as answer]
I don't think we have special code to expand `~`? May I ask why you think we have? Most of the magic we have regarding handling of paths is likely implemented by Rust libstd.

**Reply by sharkdp** [reply]
I have experienced in the past that it is not well known that `~` is *not* some special path on Linux. Or something that every Linux application supports. `~` is a shell feature. If you run `echo ~`, you will see that your shell expands `~` to your home directory path. So if someone calls `bat ~/file`, `bat` will never see that `~` character. It does not need special handling for the `~` symbol.

The question is probably hinting at what happens if `bat` actually sees that `~` symbol (?). Like if you call `bat '~/file'`. The answer is that this will lead to a "No such file or directory" error. And that's the only reasonable thing to do (and the exact same thing that other Unix applications would also do). Because there *could* actually be a folder called `~`, with a file called `file` inside. Again: there is nothing special about the `~` symbol when it comes to Linux paths.

I guess that *some* applications probably have a [`expanduser`](https://docs.python.org/3/library/os.path.html#os.path.expanduser) feature where they would *also* resolve a literal `~/file` to `/home/xxx/file`. But I don't think that's good practice.

**Reply by Enselic** [reply]
Yeah, the way this played out in my head was that @kirusfg was going to say "but why does `bat ~/file.rs` work with bat then? And _then_ I would explain what you explained above :) But I wouldn't have done such a stellar job of it as you did.

You probably want this as a top-level answer directed at @kirusfg though, right?

If not: @kirusfg see above

**Reply by kirusfg** [reply]
Thanks for your exhaustive answers! 

I did **not** know that `~` is a shell-only feature, and the problem I have is not in expanding this character when getting the path from the shell, but rather when parsing a configuration file.

Crates like [`dirs`](https://crates.io/crates/dirs) and [`shellexpand`](https://crates.io/crates/shellexpand) seem to offer what I need.

Again, thanks for answering despite this having very little to do with `bat` itself.
